src/elastio-snap.c

```diff
@@ -3149,6 +3151,18 @@ static void on_bio_read_complete(struct bio *bio){
 }
 #endif
 
+static int memory_is_too_low(struct snap_device *dev) {
+	int ret;
+	struct sysinfo si;
+	si_meminfo(&si);
+	ret = ((si.freeram * 100) / si.totalram) < LOW_MEMORY_FAIL_PERCENT ? -ENOMEM : 0;
+	if (ret) {
+		LOG_ERROR(ret, "memory is too low");
```

Comment:
I assume this message will appear in the `dmesg` output?  It's rather unsufficient by itself.

This message should include:
* What is the definition of "too low"?  (I think it's less than 10% of total physical RAM available as free memory but I'm not sure if `totalram` includes page file or not)
* What is the consequence of this error?  Will the snapshot device fail?  Is changed block tracking no longer active?  Will this cause data corruption?
* Is there anything the user needs to do to clean up after this error, or will it all happen automatically?  Could the next backup succeed if memory usage goes back down?

And not related to the message text, but a question from me:

If the memory usage is spiking *because* of `elastio-snap` recording changes to the COW file, will the failure caused by this function result in that memory being freed?

Reply:
It's 10% of total physical RAM. without counting page file. Perform querying the page file free space is a slower than querying just a free RAM so I prefer to consider only the RAM size.

Snapshot will be failed by setting it's state to respective failed value. So user space is always able to send IOCTL_ELASTIO_SNAP_INFO to get that state.
When snapshot is failed by any reason it's data no longer is consistent and of course is corrupted.

When memory overflow is caused by the driver itself and 10% limit trigger a failed state, then all the allocated by driver memory is going to be freed.

Reply:
user space side flow:

if ( IOCTL_ELASTIO_SNAP_INFO == failed_state ) {
   send IOCTL_DESTROY to destroy current failed snapshot;
   send IOCTL_SETUP_SNAP to create a new snapshot;
}

Reply:
@anelson Regarding to all details, could you suggest a meaningful but short error message for dmesg?

Reply:
@alexxxsereda or @vsazhenyuk-softheme : is that userspace flow implemented in the CLI today, or does it need to be modified to handle this failure case correctly?

Reply:
@alexxxsereda I hope that this calculation will not consider memory used by cache to be "used", since Linux systems lazily purge in-memory cache when there is no memory pressure.  If that's not the case, then this logic won't work well and will cause snapshots to fail unnecessarily.

Assuming this doesn't take into account RAM used for cache, then here's my attempt at a message:


```
Free RAM is $FREE_RAM bytes, which is under 10% of total physical memory.  Changed block tracking is unavailable when free memory falls below this threshold.  Backups can resume when more physical memory is freed.  Note that this may happen during backups under heavy write load if the underlying block device is slower than a typical SSD (for example, AWS EBS gp2/gp3 disks are known to have this problem).
```

Reply:
@alexxxsereda @anelson I think such a message is enough
```
LOG_ERROR(ret, "physical memory usage has exceeded %d%% threshold. entering error state", (100 - LOW_MEMORY_FAIL_PERCENT));
```

Reply:
> @alexxxsereda or @vsazhenyuk-softheme : is that userspace flow implemented in the CLI today, or does it need to be modified to handle this failure case correctly?

@anelson  AFAIK the CLI snap flow, after IOCTL_SETUP_SNAP the should be the next ones
IOCTL_TRANSITION_SNAP
IOCTL_TRANSITION_INC
and they should fail as fail state is on for that snapshot device.

Reply:
> @alexxxsereda @anelson I think such a message is enough
> 
> ```
> LOG_ERROR(ret, "physical memory usage has exceeded %d%% threshold. entering error state", (100 - LOW_MEMORY_FAIL_PERCENT));
> ```

Looks good for me 👍

Reply:
> @alexxxsereda I hope that this calculation will not consider memory used by cache to be "used", since Linux systems lazily purge in-memory cache when there is no memory pressure. If that's not the case, then this logic won't work well and will cause snapshots to fail unnecessarily.
> 
> Assuming this doesn't take into account RAM used for cache, then here's my attempt at a message:
> 
> ```
> Free RAM is $FREE_RAM bytes, which is under 10% of total physical memory.  Changed block tracking is unavailable when free memory falls below this threshold.  Backups can resume when more physical memory is freed.  Note that this may happen during backups under heavy write load if the underlying block device is slower than a typical SSD (for example, AWS EBS gp2/gp3 disks are known to have this problem).
> ```

@anelson Maybe it would be reasonable to have such a long message for CLI side? I like @e-kov message line as it looks much better as for dmesg.

Reply:
Okay, who will modify the CLI to handle this `E_NOMEM` error properly?

And what about this comment:

> I hope that this calculation will not consider memory used by cache to be "used", since Linux systems lazily purge in-memory cache when there is no memory pressure. If that's not the case, then this logic won't work well and will cause snapshots to fail unnecessarily.

@alexxxsereda can you confirm that cached memory will not be counted against "free" memory for the purposes of this calculation?

Reply:
ubuntu@ip-172-31-1-235:~$ cat /proc/meminfo
MemTotal:       16384740 kB
**MemFree:        13792260 kB**
MemAvailable:   15888352 kB
Buffers:           68860 kB
Cached:          2232104 kB
SwapCached:            0 kB
Active:           550984 kB
Inactive:        1828292 kB
....

Reply:
> Okay, who will modify the CLI to handle this `E_NOMEM` error properly?

@JusSergey 
https://github.com/elastio/elastio/issues/4940